
1.   카카오톡 데마운동에 올라온 **association_rules.csv**를 사용했음.

2.   이중에서 왼쪽이 5개이고 오른쪽이 1개인걸 6개로 묶어서 사용함. (지지도, 컨피던스, 이런 정보는 사라져있음) 

3. 최종적으로 뽑혀진 29개의 중복되지 않는 ard를 **최종_ard.csv**로 저장함. 

4. 원본  **main_exercises.csv**에서 결측치가 99프로 이상 존재하는 열을 제거했고, 다시 거기에서 각 열에 대해서 결측치가 존재하는 걸 삭제함.

5. 최종_ard.csv와 비교하여 최종_ard의 0번째 행이 0번째 exercises에 존재하면 0, 1번째 행이 0번째 exercises에 존재하면 1, ... [0,1] 이런식으로 ***'라벨링'***이라는 칼럼을 뒤에 새로 생성하고 리스트 저장함.

6. 직관적으로 이해할 수 있게 차원이 조금 커졌지만 더미변수화해서 각 행에 0번째 운동이 존재하면 1, 존재하지 않으면 0, ..., 28번째 운동이 존재하면 1, 존재하지 않으면 0. 이런식으로도 저장해놨음. 칼럼명은 **"_라벨링_더미0,...,라벨링_더미28_"**임 

7. 최종 결과를 **main_exercises_ard.csv**로 저장해놓음.




In [53]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [54]:
import pandas as pd
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

# 파일 경로 설정
file_path = '/content/drive/MyDrive/데이터마이닝/association_rules.csv'

# CSV 파일 불러오기
df = pd.read_csv(file_path)


In [55]:
def all_count_left_exercises(data):
    exercise_counts = {}

    for row in data:
        exercises = row.split("=>")[0].strip().split(",")
        left_count = len(exercises)
        if left_count not in exercise_counts:
            exercise_counts[left_count] = 1
        else:
            exercise_counts[left_count] += 1

    return exercise_counts


result = all_count_left_exercises(df['rules'])

for count, frequency in result.items():
    print(f"왼쪽 운동 {count}개인 것은 총 {frequency}개 있습니다.")


왼쪽 운동 1개인 것은 총 16개 있습니다.
왼쪽 운동 2개인 것은 총 376개 있습니다.
왼쪽 운동 3개인 것은 총 754개 있습니다.
왼쪽 운동 4개인 것은 총 511개 있습니다.
왼쪽 운동 5개인 것은 총 147개 있습니다.
왼쪽 운동 6개인 것은 총 20개 있습니다.


#우측이 2개이상으로 나온게 있는지 살펴봄

In [56]:
def count_left_exercises(data):
    exercise_counts = {}

    for i, row in enumerate(data):
        exercises = row.split("=>")[1].strip().split(",")
        right_count = len(exercises)
        if right_count not in exercise_counts:
            exercise_counts[right_count] = [i]
        else:
            exercise_counts[right_count].append(i)

    return exercise_counts


result = count_left_exercises(df['rules'])
line_count = 0  # 출력된 줄 개수를 저장하는 변수

for count, indices in result.items():
    if count == 2:
        for index in indices:
            print(df['rules'][index])
            line_count += 1

print(f"총 {line_count}줄이 출력되었습니다.")


총 0줄이 출력되었습니다.


#왼쪽운동이 2개이고 오른쪽 운동이 1개인 경우

In [57]:
def count_left_exercises(data):
    exercise_counts = {}

    for i, row in enumerate(data):
        exercises = row.split("=>")[0].strip().split(",")
        left_count = len(exercises)
        if left_count not in exercise_counts:
            exercise_counts[left_count] = [i]
        else:
            exercise_counts[left_count].append(i)

    return exercise_counts


result = count_left_exercises(df['rules'])
line_count = 0  # 출력된 줄 개수를 저장하는 변수

for count, indices in result.items():
    if count == 2:
        for index in indices:
            print(df['rules'][index])
            line_count += 1

print(f"총 {line_count}줄이 출력되었습니다.")


{벽에서팔굽혀펴기,엎드려양팔및다리들어올리기} => {몸통들어올리기}
{몸통들어올리기,바로서서상체숙이기} => {팔굽혀펴기}
{앉아서뒤로당기기,앉아서밀기} => {앉아서당겨내리기}
{몸통들어올리기,물통으로양팔들어올리기} => {벽에서팔굽혀펴기}
{넙다리앞쪽스트레칭,배스트레칭} => {넙다리안쪽스트레칭}
{넙다리안쪽스트레칭,배스트레칭} => {넙다리앞쪽스트레칭}
{넙다리앞쪽스트레칭,배스트레칭} => {허리스트레칭}
{넙다리앞쪽스트레칭,배스트레칭} => {넙다리뒤쪽스트레칭}
{넙다리앞쪽스트레칭,옆구리스트레칭} => {배스트레칭}
{넙다리앞쪽스트레칭,허리스트레칭} => {넙다리안쪽스트레칭}
{넙다리안쪽스트레칭,허리스트레칭} => {넙다리앞쪽스트레칭}
{넙다리안쪽스트레칭,넙다리앞쪽스트레칭} => {넙다리뒤쪽스트레칭}
{넙다리뒤쪽스트레칭,넙다리앞쪽스트레칭} => {넙다리안쪽스트레칭}
{넙다리앞쪽스트레칭,옆구리스트레칭} => {넙다리안쪽스트레칭}
{넙다리안쪽스트레칭,옆구리스트레칭} => {넙다리앞쪽스트레칭}
{넙다리앞쪽스트레칭,허리스트레칭} => {넙다리뒤쪽스트레칭}
{넙다리앞쪽스트레칭,옆구리스트레칭} => {허리스트레칭}
{넙다리앞쪽스트레칭,옆구리스트레칭} => {넙다리뒤쪽스트레칭}
{누워서다리좌우로움직이기,윗몸올리기} => {엎드려버티기}
{앉아서다리굽히기,앉아서다리밀기} => {앉아서다리펴기}
{엎드려팔다리교차올리기,윗몸일으키기} => {전완대고버티기}
{누워서팔다리동시에들어올리기,윗몸올리기} => {엎드려버티기}
{뒤꿈치높이들어뛰기,버피테스트} => {순간반응콘찍기}
{버피테스트,순간반응콘찍기} => {뒤꿈치높이들어뛰기}
{뒤꿈치높이들어뛰기,순간반응콘찍기} => {맨몸운동루틴프로그램}
{뒤꿈치높이들어뛰기,맨몸운동루틴프로그램} => {순간반응콘찍기}
{맨몸운동루틴프로그램,순간반응콘찍기} => {뒤꿈치높이들어뛰기}
{뒤꿈치높이들어뛰기,버피테스트} => {맨몸운동루틴프로그램}
{뒤꿈치높이들어뛰기,맨몸운동루틴프로그램} => {버피테스트}
{발바꿔뛰기,크런치사이클} => {

In [58]:
from collections import Counter

def count_left_exercises(data):
    left_exercises_count = []

    for row in data:
        row = row.replace("{", "").replace("}", "")  # "{"와 "}" 문자열 제거
        left_exercises, right_exercises = row.split("=>")
        left_exercises = left_exercises.strip().split(",")
        right_exercise = right_exercises.strip()

        if len(left_exercises) == 2:  # 왼쪽 운동이 2개인 경우
            combination = tuple(sorted(left_exercises + [right_exercise]))
            left_exercises_count.append(combination)

    return Counter(left_exercises_count)


result = count_left_exercises(df['rules'])
total_lines = len(result)


for combination, frequency in result.items():
    print(f"{', '.join(combination)} (빈도: {frequency})")
    
print("총",total_lines,"줄이 출력되었습니다.")

몸통들어올리기, 벽에서팔굽혀펴기, 엎드려양팔및다리들어올리기 (빈도: 1)
몸통들어올리기, 바로서서상체숙이기, 팔굽혀펴기 (빈도: 1)
앉아서당겨내리기, 앉아서뒤로당기기, 앉아서밀기 (빈도: 1)
몸통들어올리기, 물통으로양팔들어올리기, 벽에서팔굽혀펴기 (빈도: 1)
넙다리안쪽스트레칭, 넙다리앞쪽스트레칭, 배스트레칭 (빈도: 2)
넙다리앞쪽스트레칭, 배스트레칭, 허리스트레칭 (빈도: 1)
넙다리뒤쪽스트레칭, 넙다리앞쪽스트레칭, 배스트레칭 (빈도: 1)
넙다리앞쪽스트레칭, 배스트레칭, 옆구리스트레칭 (빈도: 1)
넙다리안쪽스트레칭, 넙다리앞쪽스트레칭, 허리스트레칭 (빈도: 2)
넙다리뒤쪽스트레칭, 넙다리안쪽스트레칭, 넙다리앞쪽스트레칭 (빈도: 2)
넙다리안쪽스트레칭, 넙다리앞쪽스트레칭, 옆구리스트레칭 (빈도: 2)
넙다리뒤쪽스트레칭, 넙다리앞쪽스트레칭, 허리스트레칭 (빈도: 1)
넙다리앞쪽스트레칭, 옆구리스트레칭, 허리스트레칭 (빈도: 1)
넙다리뒤쪽스트레칭, 넙다리앞쪽스트레칭, 옆구리스트레칭 (빈도: 1)
누워서다리좌우로움직이기, 엎드려버티기, 윗몸올리기 (빈도: 1)
앉아서다리굽히기, 앉아서다리밀기, 앉아서다리펴기 (빈도: 1)
엎드려팔다리교차올리기, 윗몸일으키기, 전완대고버티기 (빈도: 1)
누워서팔다리동시에들어올리기, 엎드려버티기, 윗몸올리기 (빈도: 1)
뒤꿈치높이들어뛰기, 버피테스트, 순간반응콘찍기 (빈도: 2)
뒤꿈치높이들어뛰기, 맨몸운동루틴프로그램, 순간반응콘찍기 (빈도: 3)
뒤꿈치높이들어뛰기, 맨몸운동루틴프로그램, 버피테스트 (빈도: 2)
발바꿔뛰기, 버피테스트, 크런치사이클 (빈도: 1)
몸통들어올리기, 버피테스트, 크런치사이클 (빈도: 1)
버피테스트, 크런치사이클, 팔굽혀펴기 (빈도: 1)
버피테스트, 앉았다일어서기, 크런치사이클 (빈도: 1)
앉았다일어서기, 크런치사이클, 팔굽혀펴기 (빈도: 1)
넙다리안쪽스트레칭, 배스트레칭, 허리스트레칭 (빈도: 1)
넙다리뒤쪽스트레칭, 넙다리안쪽스트레칭, 배스트레칭 (빈도: 1)
배스트레칭, 엉덩이스트

In [59]:
def count_combinations_with_frequency_above_2(data):
    result = count_left_exercises(data)
    count = 0

    for frequency in result.values():
        if frequency == 2:
            count += 1

    return count


frequency_above_2_count = count_combinations_with_frequency_above_2(df['rules'])
print("빈도가 2 인 조합의 개수:", frequency_above_2_count)


빈도가 2 인 조합의 개수: 59


In [60]:
def count_combinations_with_frequency_above_3(data):
    result = count_left_exercises(data)
    count = 0

    for frequency in result.values():
        if frequency >= 3:
            count += 1

    return count


frequency_above_3_count = count_combinations_with_frequency_above_3(df['rules'])
print("빈도가 3 인 조합의 개수:", frequency_above_3_count)


빈도가 3 인 조합의 개수: 1


1곱하기2+59곱하기1+315=376줄 검산맞음

#왼쪽운동이 3개이고 오른쪽 운동이 1개인 경우

In [61]:
from collections import Counter

def count_left_exercises(data):
    left_exercises_count = []

    for row in data:
        row = row.replace("{", "").replace("}", "")  # "{"와 "}" 문자열 제거
        left_exercises, right_exercises = row.split("=>")
        left_exercises = left_exercises.strip().split(",")
        right_exercise = right_exercises.strip()

        if len(left_exercises) == 3:  # 왼쪽 운동이 3개인 경우
            combination = tuple(sorted(left_exercises + [right_exercise]))
            left_exercises_count.append(combination)

    return Counter(left_exercises_count)


result = count_left_exercises(df['rules'])
total_lines = len(result)


for combination, frequency in result.items():
    print(f"{', '.join(combination)} (빈도: {frequency})")
    
print("총", total_lines, "줄이 출력되었습니다.")


넙다리안쪽스트레칭, 넙다리앞쪽스트레칭, 배스트레칭, 허리스트레칭 (빈도: 3)
넙다리뒤쪽스트레칭, 넙다리안쪽스트레칭, 넙다리앞쪽스트레칭, 배스트레칭 (빈도: 3)
넙다리뒤쪽스트레칭, 넙다리앞쪽스트레칭, 배스트레칭, 허리스트레칭 (빈도: 3)
넙다리앞쪽스트레칭, 배스트레칭, 옆구리스트레칭, 허리스트레칭 (빈도: 2)
넙다리뒤쪽스트레칭, 넙다리앞쪽스트레칭, 배스트레칭, 옆구리스트레칭 (빈도: 4)
넙다리뒤쪽스트레칭, 넙다리안쪽스트레칭, 넙다리앞쪽스트레칭, 허리스트레칭 (빈도: 3)
넙다리안쪽스트레칭, 넙다리앞쪽스트레칭, 옆구리스트레칭, 허리스트레칭 (빈도: 3)
넙다리뒤쪽스트레칭, 넙다리안쪽스트레칭, 넙다리앞쪽스트레칭, 엉덩이스트레칭 (빈도: 2)
넙다리뒤쪽스트레칭, 넙다리안쪽스트레칭, 넙다리앞쪽스트레칭, 옆구리스트레칭 (빈도: 3)
넙다리뒤쪽스트레칭, 넙다리앞쪽스트레칭, 옆구리스트레칭, 허리스트레칭 (빈도: 3)
뒤꿈치높이들어뛰기, 맨몸운동루틴프로그램, 버피테스트, 순간반응콘찍기 (빈도: 4)
넙다리뒤쪽스트레칭, 넙다리안쪽스트레칭, 배스트레칭, 허리스트레칭 (빈도: 3)
넙다리뒤쪽스트레칭, 넙다리안쪽스트레칭, 배스트레칭, 옆구리스트레칭 (빈도: 1)
넙다리뒤쪽스트레칭, 배스트레칭, 엉덩이스트레칭, 허리스트레칭 (빈도: 1)
배스트레칭, 엉덩이스트레칭, 옆구리스트레칭, 허리스트레칭 (빈도: 2)
넙다리뒤쪽스트레칭, 배스트레칭, 옆구리스트레칭, 허리스트레칭 (빈도: 2)
넙다리뒤쪽스트레칭, 넙다리안쪽스트레칭, 옆구리스트레칭, 허리스트레칭 (빈도: 2)
무릎높이들어뛰기, 자전거타기, 조깅, 줄넘기운동 (빈도: 3)
물병옆으로들어올리기, 팔꿈치굽히기, 한발앞으로내밀고앉았다일어서기, 허리굽혀팔뒤로들기 (빈도: 2)
물병옆으로들어올리기, 앉았다일어서기, 팔꿈치굽히기, 허리굽혀팔뒤로들기 (빈도: 1)
물병옆으로들어올리기, 앉았다일어서기, 팔꿈치굽히기, 한발앞으로내밀고앉았다일어서기 (빈도: 3)
물병옆으로들어올리기, 앉았다일어서기, 윗몸올리기, 팔꿈치굽히기 (빈도: 3)
물병

In [62]:
def count_combinations_with_frequency_above_2(data):
    result = count_left_exercises(data)
    count = 0

    for frequency in result.values():
        if frequency == 2:
            count += 1

    return count


frequency_above_2_count = count_combinations_with_frequency_above_2(df['rules'])
print("빈도가 2 인 조합의 개수:", frequency_above_2_count)


빈도가 2 인 조합의 개수: 147


In [63]:
def count_combinations_with_frequency_above_3(data):
    result = count_left_exercises(data)
    count = 0

    for frequency in result.values():
        if frequency == 3:
            count += 1

    return count


frequency_above_3_count = count_combinations_with_frequency_above_3(df['rules'])
print("빈도가 3 인 조합의 개수:", frequency_above_3_count)


빈도가 3 인 조합의 개수: 97


In [64]:
def count_combinations_with_frequency_above_4(data):
    result = count_left_exercises(data)
    count = 0

    for frequency in result.values():
        if frequency == 4:
            count += 1

    return count


frequency_above_4_count = count_combinations_with_frequency_above_4(df['rules'])
print("빈도가 4 인 조합의 개수:", frequency_above_4_count)


빈도가 4 인 조합의 개수: 17


In [65]:
def count_combinations_with_frequency_above_5(data):
    result = count_left_exercises(data)
    count = 0

    for frequency in result.values():
        if frequency > 4:
            count += 1

    return count


frequency_above_5_count = count_combinations_with_frequency_above_5(df['rules'])
print("빈도가 4 초과인 조합의 개수:", frequency_above_5_count)


빈도가 4 초과인 조합의 개수: 0


17곱하기3+97곱하기2+147곱하기1+362 =754줄 검산 맞음


In [66]:
from collections import Counter

def count_left_exercises(data):
    left_exercises_count = []

    for row in data:
        row = row.replace("{", "").replace("}", "")  # "{"와 "}" 문자열 제거
        left_exercises, right_exercises = row.split("=>")
        left_exercises = left_exercises.strip().split(",")
        right_exercise = right_exercises.strip()

        if len(left_exercises) == 4:  # 왼쪽 운동이 3개인 경우
            combination = tuple(sorted(left_exercises + [right_exercise]))
            left_exercises_count.append(combination)

    return Counter(left_exercises_count)


result = count_left_exercises(df['rules'])
total_lines = len(result)


for combination, frequency in result.items():
    print(f"{', '.join(combination)} (빈도: {frequency})")
    
print("총", total_lines, "줄이 출력되었습니다.")


넙다리뒤쪽스트레칭, 넙다리안쪽스트레칭, 넙다리앞쪽스트레칭, 배스트레칭, 허리스트레칭 (빈도: 5)
넙다리뒤쪽스트레칭, 넙다리앞쪽스트레칭, 배스트레칭, 옆구리스트레칭, 허리스트레칭 (빈도: 5)
앉았다일어서기, 윗몸올리기, 팔꿈치굽히기, 한발앞으로내밀고앉았다일어서기, 허리굽혀팔뒤로들기 (빈도: 5)
계단뛰어오르기, 달리기, 실외자전거타기, 줄넘기, 트레드밀에서걷기 (빈도: 3)
계단뛰어오르기, 계단올라갔다내려오기, 실내자전거타기, 줄넘기, 트레드밀에서걷기 (빈도: 4)
계단뛰어오르기, 계단올라갔다내려오기, 달리기, 수영, 실내자전거타기 (빈도: 2)
서서상체일으키기, 앉아서다리밀기, 앉아서다리펴기, 앉아서당겨내리기, 앉아서밀기 (빈도: 4)
서서상체일으키기, 앉아서다리밀기, 앉아서다리펴기, 앉아서밀기, 윗몸올리기 (빈도: 5)
서서상체일으키기, 앉아서다리펴기, 앉아서당겨내리기, 앉아서밀기, 윗몸올리기 (빈도: 5)
서서상체일으키기, 앉아서다리밀기, 앉아서당겨내리기, 앉아서밀기, 윗몸올리기 (빈도: 5)
실내자전거타기, 앉아서다리밀기, 앉아서다리펴기, 앉아서당겨내리기, 앉아서밀기 (빈도: 4)
앉아서다리밀기, 앉아서다리펴기, 앉아서당겨내리기, 앉아서밀기, 윗몸올리기 (빈도: 4)
앉아서다리밀기, 앉아서다리펴기, 앉아서당겨내리기, 앉아서밀기, 엎드려버티기 (빈도: 4)
누워서엉덩이들어올리기, 앉아서다리밀기, 앉아서다리펴기, 앉아서밀기, 엎드려버티기 (빈도: 4)
실내자전거타기, 앉아서다리밀기, 앉아서다리펴기, 앉아서밀기, 윗몸올리기 (빈도: 4)
앉아서다리밀기, 앉아서다리펴기, 앉아서밀기, 엎드려버티기, 윗몸올리기 (빈도: 4)
실내자전거타기, 앉아서다리펴기, 앉아서당겨내리기, 앉아서밀기, 윗몸올리기 (빈도: 5)
앉아서다리펴기, 앉아서당겨내리기, 앉아서밀기, 엎드려버티기, 윗몸올리기 (빈도: 4)
실내자전거타기, 앉아서다리밀기, 앉아서당겨내리기, 앉아서밀기, 윗몸올리기 (빈도: 5)
앉아서다리밀기, 앉아서당겨내리기, 앉아서밀기, 엎드려버티기, 윗몸올리기 (빈도: 

In [67]:
from collections import Counter

def count_left_exercises(data):
    left_exercises_count = []

    for row in data:
        row = row.replace("{", "").replace("}", "")  # "{"와 "}" 문자열 제거
        left_exercises, right_exercises = row.split("=>")
        left_exercises = left_exercises.strip().split(",")
        right_exercise = right_exercises.strip()

        if len(left_exercises) == 5:  # 왼쪽 운동이 3개인 경우
            combination = tuple(sorted(left_exercises + [right_exercise]))
            left_exercises_count.append(combination)

    return Counter(left_exercises_count)


result = count_left_exercises(df['rules'])
total_lines = len(result)


for combination, frequency in result.items():
    print(f"{', '.join(combination)} (빈도: {frequency})")
    
print("총", total_lines, "줄이 출력되었습니다.")


서서상체일으키기, 앉아서다리밀기, 앉아서다리펴기, 앉아서당겨내리기, 앉아서밀기, 윗몸올리기 (빈도: 6)
앉아서다리밀기, 앉아서다리펴기, 앉아서당겨내리기, 앉아서밀기, 엎드려버티기, 윗몸올리기 (빈도: 6)
누워서엉덩이들어올리기, 실내자전거타기, 앉아서다리밀기, 앉아서다리펴기, 엎드려버티기, 윗몸올리기 (빈도: 6)
누워서엉덩이들어올리기, 실내자전거타기, 앉아서다리밀기, 앉아서당겨내리기, 윗몸올리기, 트레드밀에서걷기 (빈도: 6)
누워서엉덩이들어올리기, 실내자전거타기, 앉아서다리밀기, 앉아서당겨내리기, 엎드려버티기, 트레드밀에서걷기 (빈도: 6)
누워서엉덩이들어올리기, 앉아서다리밀기, 앉아서당겨내리기, 엎드려버티기, 윗몸올리기, 트레드밀에서걷기 (빈도: 6)
실내자전거타기, 앉아서다리밀기, 앉아서당겨내리기, 엎드려버티기, 윗몸올리기, 트레드밀에서걷기 (빈도: 6)
누워서엉덩이들어올리기, 실내자전거타기, 앉아서다리밀기, 앉아서당겨내리기, 엎드려버티기, 윗몸올리기 (빈도: 6)
누워서엉덩이들어올리기, 실내자전거타기, 앉아서다리밀기, 엎드려버티기, 윗몸올리기, 트레드밀에서걷기 (빈도: 4)
누워서다리들어올리기, 앉았다일어서기, 엎드려버티기, 옆으로누워버티기, 윗몸올리기, 팔굽혀펴기 (빈도: 5)
누워서엉덩이들어올리기, 실내자전거타기, 앉아서당겨내리기, 엎드려버티기, 윗몸올리기, 트레드밀에서걷기 (빈도: 5)
누워서다리들어올리기, 앉았다일어서기, 엎드려버티기, 엎드려서팔다리들기, 윗몸올리기, 팔굽혀펴기 (빈도: 3)
걷기, 수영, 자전거타기, 전완대고버티기, 조깅, 줄넘기운동 (빈도: 4)
걷기, 앉았다일어서기, 자전거타기, 전완대고버티기, 조깅, 줄넘기운동 (빈도: 6)
걷기, 수영, 앉았다일어서기, 자전거타기, 전완대고버티기, 줄넘기운동 (빈도: 6)
걷기, 수영, 앉았다일어서기, 자전거타기, 전완대고버티기, 조깅 (빈도: 6)
수영, 앉았다일어서기, 자전거타기, 전완대고버티기, 조깅, 줄넘기운동 (빈도: 4)
걷기, 수영, 앉았다일어서기, 전완대고버티기, 조깅,

In [68]:
import csv

def count_left_exercises(data):
    left_exercises_count = []

    for row in data:
        row = row.replace("{", "").replace("}", "")  # "{"와 "}" 문자열 제거
        left_exercises, right_exercises = row.split("=>")
        left_exercises = left_exercises.strip().split(",")
        right_exercise = right_exercises.strip()

        if len(left_exercises) == 5:  # 왼쪽 운동이 5개인 경우
            combination = tuple(sorted(left_exercises + [right_exercise]))
            left_exercises_count.append(combination)

    return left_exercises_count


result = count_left_exercises(df['rules'])
unique_lines = set(result)  # 중복 행 제거

output_path = "/content/drive/MyDrive/데이터마이닝/데이터셋/최종_ard.csv"  # 원하는 출력 경로와 파일명 설정

with open(output_path, 'w', newline='') as csvfile:
    writer = csv.writer(csvfile)
    writer.writerow(["최종_ard"])

    for line in unique_lines:
        writer.writerow([line])


In [69]:
# 파일 경로 설정
file_path = ('/content/drive/MyDrive/데이터마이닝/데이터셋/최종_ard.csv')

# CSV 파일 불러오기
df_ard = pd.read_csv(file_path)


In [70]:
import re

df_ard['최종_ard'] = df_ard['최종_ard'].apply(lambda x: set(re.findall(r'\w+', x)))


In [71]:
print(df_ard)

                                               최종_ard
0   {누워서엉덩이들어올리기, 엎드려버티기, 앉아서다리펴기, 실내자전거타기, 앉아서다리밀...
1   {앉아서다리펴기, 엎드려버티기, 앉아서다리밀기, 윗몸올리기, 앉아서밀기, 앉아서당겨...
2            {엎드려버티기, 윗몸올리기, 줄넘기, 팔굽혀펴기, 수영, 앉았다일어서기}
3   {엎드려버티기, 트레드밀에서걷기, 실내자전거타기, 앉아서다리밀기, 윗몸올리기, 앉아...
4              {엎드려버티기, 줄넘기, 팔굽혀펴기, 수영, 앉았다일어서기, 달리기}
5   {누워서엉덩이들어올리기, 엎드려버티기, 트레드밀에서걷기, 실내자전거타기, 윗몸올리기...
6                {엎드려버티기, 윗몸올리기, 줄넘기, 팔굽혀펴기, 수영, 달리기}
7   {누워서엉덩이들어올리기, 엎드려서팔다리들기, 윗몸올리기, 팔굽혀펴기, 한발앞으로내밀...
8   {누워서엉덩이들어올리기, 트레드밀에서걷기, 실내자전거타기, 앉아서다리밀기, 윗몸올리...
9   {누워서엉덩이들어올리기, 엎드려버티기, 트레드밀에서걷기, 앉아서다리밀기, 윗몸올리기...
10              {걷기, 줄넘기운동, 조깅, 수영, 앉았다일어서기, 전완대고버티기}
11              {걷기, 자전거타기, 조깅, 수영, 앉았다일어서기, 전완대고버티기}
12           {줄넘기운동, 자전거타기, 조깅, 수영, 앉았다일어서기, 전완대고버티기}
13        {맨몸운동루틴프로그램, 줄넘기운동, 자전거타기, 조깅, 수영, 앉았다일어서기}
14           {걷기, 줄넘기운동, 자전거타기, 조깅, 앉았다일어서기, 전완대고버티기}
15             {엎드려버티기, 윗몸올리기, 줄넘기, 수영, 앉았다일어서기, 달리기}
16  {누워서엉덩이들어올리기, 엎드려버티기, 트레드밀에서걷기, 실내자전거타기, 앉아서다리...
17  {누워서엉덩이들어올리기, 엎드려버티기, 엎드

In [72]:
import pandas as pd
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

# 파일 경로 설정
file_path = ('/content/drive/MyDrive/데이터마이닝/데이터셋/main_exercises.csv')

# CSV 파일 불러오기
df = pd.read_csv(file_path)

# exercises가 [''] 값을 가지는 행 제거
df = df[df['exercises'] != "['']"]

#각 행에 대한 결측치 비율
empty_ratio = df.isnull().mean() * 100

#해당 열의 결측치 비율이 99프로가 넘어가는거 싹다 지우기
columns_to_drop = empty_ratio[empty_ratio > 1].index
df = df.drop(columns=columns_to_drop)

#결측 열을 날린 데이터셋에 대하여 각 행에 남아있는 결측치 지우기
for column in df.columns:
    if df[column].isnull().any():
        df = df.dropna(subset=[column])

In [73]:
#살아남은 어트리뷰트
for column_name in df.columns:
    print(column_name)

Unnamed: 0
MBER_SEQ_NO_VALUE
MESURE_SEQ_NO
CNTER_NM
AGRDE_FLAG_NM
MESURE_PLACE_FLAG_NM
MESURE_AGE_CO
INPT_FLAG_NM
CRTFC_FLAG_NM
MESURE_DE
SEXDSTN_FLAG_CD
MESURE_IEM_001_VALUE
MESURE_IEM_002_VALUE
MESURE_IEM_005_VALUE
MESURE_IEM_006_VALUE
MESURE_IEM_007_VALUE
MESURE_IEM_008_VALUE
MESURE_IEM_012_VALUE
MESURE_IEM_018_VALUE
MESURE_IEM_028_VALUE
MVM_PRSCRPTN_CN
exercises


In [74]:
#각 행에 대한 결측치 비율 다시한번 확인
empty_ratio = df.isnull().mean() * 100
print(empty_ratio)

Unnamed: 0              0.0
MBER_SEQ_NO_VALUE       0.0
MESURE_SEQ_NO           0.0
CNTER_NM                0.0
AGRDE_FLAG_NM           0.0
MESURE_PLACE_FLAG_NM    0.0
MESURE_AGE_CO           0.0
INPT_FLAG_NM            0.0
CRTFC_FLAG_NM           0.0
MESURE_DE               0.0
SEXDSTN_FLAG_CD         0.0
MESURE_IEM_001_VALUE    0.0
MESURE_IEM_002_VALUE    0.0
MESURE_IEM_005_VALUE    0.0
MESURE_IEM_006_VALUE    0.0
MESURE_IEM_007_VALUE    0.0
MESURE_IEM_008_VALUE    0.0
MESURE_IEM_012_VALUE    0.0
MESURE_IEM_018_VALUE    0.0
MESURE_IEM_028_VALUE    0.0
MVM_PRSCRPTN_CN         0.0
exercises               0.0
dtype: float64


In [75]:
df

Unnamed: 0         MBER_SEQ_NO_VALUE  MESURE_SEQ_NO CNTER_NM  \
15               15  AAFAqwiBa9l7F9QH7QjeCnUD              1       광명   
16               16  AAHcEr8ziYVuy1vI090rFlfd              1   남구(부산)   
18               18  AAF6IFD08aN3Yk7f0b31phqy              1       광명   
19               19  AAFJD2Hj3y8ufEHcjt5r3CG6              1   남구(부산)   
20               20  AAHlLv5K+dh+zZGoW8M9Z+un              1       광명   
...             ...                       ...            ...      ...   
1589306      571286  AAHCOM2hu2k2f54fWhl/WJ0x              2       청주   
1589310      571290  AAFrVuG4TA0Q6Bkd3IwrIKZ3              4       서초   
1589311      571291  AAHOBt105X3y7ckglfiNKNgO             11       진천   
1589313      571293  AAG1Kawp6edvMMPKTB5kjV7q              2       제주   
1589319      571299  AAHG1sWtEFs81x4qmbr77Wt5              2       계룡   

        AGRDE_FLAG_NM MESURE_PLACE_FLAG_NM  MESURE_AGE_CO INPT_FLAG_NM  \
15                 성인                   일반           63.0          관리자   
16                 성인                   일반           49.0          관리자   
18                 성인                   일반           62.0          관리자   
19                 성인                   일반           43.0          관리자   
20                 성인                   일반           60.0          관리자   
...               ...                  ...            ...          ...   
1589306            성인                   일반           32.0          관리자   
1589310            성인                   출장           26.0          관리자   
1589311            성인                   일반           20.0          관리자   
1589313            성인                   출장           50.0          관리자   
1589319           청소년                   일반           17.0          관리자   

        CRTFC_FLAG_NM  MESURE_DE  ... MESURE_IEM_002_VALUE  \
15                3등급   20110101  ...                63.08   
16                참가증   20110101  ...                55.02   
18                참가증   20110101  ...                43.22   
19                3등급   20110101  ...                51.42   
20                참가증   20110101  ...                94.88   
...               ...        ...  ...                  ...   
1589306           2등급   20220531  ...                72.20   
1589310           참가증   20220531  ...                84.60   
1589311           2등급   20220531  ...                60.70   
1589313           2등급   20220531  ...                69.10   
1589319           3등급   20220531  ...                73.40   

         MESURE_IEM_005_VALUE  MESURE_IEM_006_VALUE  MESURE_IEM_007_VALUE  \
15                       40.0                  97.0                  21.9   
16                       86.0                 124.0                  23.9   
18                       80.0                 120.0                  16.6   
19                       65.0                 107.0                  22.3   
20                       75.0                 116.0                  20.7   
...                       ...                   ...                   ...   
1589306                  69.0                 114.0                  46.4   
1589310                  73.0                 143.0                  42.7   
1589311                  79.0                 134.0                  40.9   
1589313                  97.0                 148.0                  36.5   
1589319                  63.0                 115.0                  43.3   

         MESURE_IEM_008_VALUE  MESURE_IEM_012_VALUE  MESURE_IEM_018_VALUE  \
15                       21.0                  20.8                  24.0   
16                       25.9                  19.3                  20.7   
18                       22.3                   9.2                  17.7   
19                       21.5                  20.7                  22.5   
20                       22.8                  15.7                  36.1   
...                       ...                   ...                   ...   
1589306                  50.7     

In [76]:
import ast
# 'exercises' 열의 문자열을 리스트 형태로 변환
df['exercises'] = df['exercises'].apply(ast.literal_eval)

In [77]:
df['exercises']

15         [옆구리스트레칭, 전신루틴스트레칭, 좌식생활자를위한동적루틴스트레칭, 계단올라갔다내려...
16         [목스트레칭, 배스트레칭, 상지루틴스트레칭, 엉덩이스트레칭, 옆구리스트레칭, 하지루...
18                                 [넙다리뒤쪽스트레칭, 넙다리앞쪽스트레칭, 등]
19         [상지루틴스트레칭, 유산소운동전동적루틴스트레칭, 전신루틴스트레칭, 하지루틴스트레칭1...
20                                                      [가슴]
                                 ...                        
1589306    [달리기, 계단뛰어오르기, 한발앞으로내밀고앉았다일어서기, 의자에한다리올리고앉앗다일어...
1589310    [의자이용유연성운동루틴프로그램, 정적스트레칭루틴프로그램, 좌식생활자를위한동적루틴스트...
1589311    [서서상체일으키기, 앉았다일어서기, 윗몸올리기, 엎드려버티기, 누워서팔다리동시에들어...
1589313    [맨몸운동루틴프로그램, 관절염을위한심폐지구력운동, 벽잡고빨리뛰기, 연속무릎올리기, ...
1589319                         [맨몸운동루틴프로그램, 전완대고버티기, 버피테스트]
Name: exercises, Length: 1194793, dtype: object

In [78]:
# 'exercises' 열의 리스트를 집합 형식으로 변환
df['exercises'] = df['exercises'].apply(lambda x: set(x))

In [79]:
df['exercises']

15         {트레드밀에서걷기, 앉아서엉덩관절굽히기, 사다리좌우뛰기, 스텝퍼오르내리기, 앉아서모...
16         {배스트레칭, 하지루틴스트레칭1, 목스트레칭, 엉덩이스트레칭, 옆구리스트레칭, 상지...
18                                 {등, 넙다리앞쪽스트레칭, 넙다리뒤쪽스트레칭}
19         {서서균형잡기, 걷기, 유산소운동전동적루틴스트레칭, 하지루틴스트레칭1, 물병옆으로들...
20                                                      {가슴}
                                 ...                        
1589306    {넙다리뒤쪽스트레칭, 의자에한다리올리고앉앗다일어서기, 점프팔굽혀펴기, 누워서배가로근...
1589310    {정적스트레칭루틴프로그램, 동적스트레칭루틴프로그램, 하지루틴스트레칭2, 넙다리뒤쪽스...
1589311    {엎드려버티기, 누워서엉덩이들어올리기, 누워서팔다리동시에들어올리기, 윗몸올리기, 팔...
1589313    {맨몸운동루틴프로그램, 짐볼앉아중심잡기루틴운동, 연속무릎올리기, 벽잡고빨리뛰기, 반...
1589319                         {전완대고버티기, 맨몸운동루틴프로그램, 버피테스트}
Name: exercises, Length: 1194793, dtype: object

In [80]:
#set으로 exercises와 최종_ard의 0번째행부터 28번째 행까지 
# 쭉 확인하면서 새로운 df['라벨링']에 리스트 형태로 운동을 저장시키는 코드
df['라벨링'] = df['exercises'].apply(lambda x: [i for i, ard in enumerate(df_ard['최종_ard']) if set(ard).issubset(x)])


In [82]:
# 나중에 분석할때 편하라고
# 최종_ard를 더미 변수화하여 라벨링 뒤에다가 라벨링_더미0, 라벨링_더미1, ... 컬럼 생성해놓은거
dummy_columns = []
for i, ard in enumerate(df_ard['최종_ard']):
    dummy_column_name = f'라벨링_더미{i}'
    df[dummy_column_name] = df['exercises'].apply(lambda x: 1 if set(ard).issubset(x) else 0)
    dummy_columns.append(dummy_column_name)


In [84]:
#일단 더미변수 빼놓고 라벨링만 출력해서 확인
print(df[['exercises', '라벨링']])

                                                 exercises  \
15       {트레드밀에서걷기, 앉아서엉덩관절굽히기, 사다리좌우뛰기, 스텝퍼오르내리기, 앉아서모...   
16       {배스트레칭, 하지루틴스트레칭1, 목스트레칭, 엉덩이스트레칭, 옆구리스트레칭, 상지...   
18                               {등, 넙다리앞쪽스트레칭, 넙다리뒤쪽스트레칭}   
19       {서서균형잡기, 걷기, 유산소운동전동적루틴스트레칭, 하지루틴스트레칭1, 물병옆으로들...   
20                                                    {가슴}   
...                                                    ...   
1589306  {넙다리뒤쪽스트레칭, 의자에한다리올리고앉앗다일어서기, 점프팔굽혀펴기, 누워서배가로근...   
1589310  {정적스트레칭루틴프로그램, 동적스트레칭루틴프로그램, 하지루틴스트레칭2, 넙다리뒤쪽스...   
1589311  {엎드려버티기, 누워서엉덩이들어올리기, 누워서팔다리동시에들어올리기, 윗몸올리기, 팔...   
1589313  {맨몸운동루틴프로그램, 짐볼앉아중심잡기루틴운동, 연속무릎올리기, 벽잡고빨리뛰기, 반...   
1589319                       {전완대고버티기, 맨몸운동루틴프로그램, 버피테스트}   

                               라벨링  
15                              []  
16                              []  
18                              []  
19                              []  
20                              []  
...                            ..

In [85]:
#라벨링 중 해당되는게 없어 [] 이런식으로 출력된거 없애는 코드
df_filtered = df[df['라벨링'].apply(lambda x: len(x) > 0)]

# 결과 출력
print(df_filtered[['exercises', '라벨링']])


                                                 exercises  \
4856     {엎드려버티기, 스텝퍼뛰어서넘어가기, 엎드려서다리차올리기, 넙다리뒤쪽스트레칭, 실외...   
4967     {엎드려버티기, 넙다리뒤쪽스트레칭, 사다리좌우뛰기, 넙다리앞쪽스트레칭, 윗몸올리기,...   
5376     {실외자전거타기, 윗몸올리기, 계단두발뛰기, 한발앞으로내밀고앉았다일어서기, 달리기,...   
5407     {엎드려서다리차올리기, 넙다리뒤쪽스트레칭, 실외자전거타기, 하지루틴스트레칭1, 윗몸...   
5409     {엎드려서다리차올리기, 넙다리뒤쪽스트레칭, 실외자전거타기, 하지루틴스트레칭1, 윗몸...   
...                                                    ...   
1589105  {엎드려버티기, 누워서수직으로다리들어올리기, 누워서하늘자전거, 누워서엉덩이들어올리기...   
1589141  {엎드려버티기, 누워서수직으로다리들어올리기, 누워서하늘자전거, 누워서엉덩이들어올리기...   
1589166  {엎드려버티기, 누워서수직으로다리들어올리기, 누워서하늘자전거, 누워서엉덩이들어올리기...   
1589196  {몸통옆으로굽히기, 엎드려버티기, 누워서하늘자전거, 서서균형잡기, 트레드밀에서걷기,...   
1589311  {엎드려버티기, 누워서엉덩이들어올리기, 누워서팔다리동시에들어올리기, 윗몸올리기, 팔...   

                               라벨링  
4856                           [4]  
4967                           [2]  
5376                          [18]  
5407                          [18]  
5409                          [18]  
...                            ..

In [86]:
#라벨링 더미까지 확인해보기 
df_filtered

Unnamed: 0         MBER_SEQ_NO_VALUE  MESURE_SEQ_NO CNTER_NM  \
4856           4856  AAEGgknbtLhqsC0Y3OBk8x8Q              1       원주   
4967           4967  AAHQGHadxmBqm97xj5UkekYM              1       광명   
5376           5376  AAGNk6oMJXD+CmWZQGAScEjF              1       원주   
5407           5407  AAHtDxjV6XKSvQFpnf7uvaSf              1       원주   
5409           5409  AAGjZxPuVaM1EkYYlBNJ7nOg              1       원주   
...             ...                       ...            ...      ...   
1589105      571085  AAFmwbuWzix3VQbgyriejbth              2       서초   
1589141      571121  AAFO8yIDc5JivYDOuxf5hIGi              2       서초   
1589166      571146  AAGjUZz1yo/6wQSndAIXWsDI              2       서초   
1589196      571176  AAH7wFF3bTe5mJzIlafQpOXz              1       계룡   
1589311      571291  AAHOBt105X3y7ckglfiNKNgO             11       진천   

        AGRDE_FLAG_NM MESURE_PLACE_FLAG_NM  MESURE_AGE_CO INPT_FLAG_NM  \
4856               성인                   일반           39.0          관리자   
4967               성인                   일반           19.0          관리자   
5376               성인                   일반           25.0          관리자   
5407               성인                   일반           19.0          관리자   
5409               성인                   일반           19.0          관리자   
...               ...                  ...            ...          ...   
1589105            성인                   출장           25.0          관리자   
1589141            성인                   출장           21.0          관리자   
1589166            성인                   출장           22.0          관리자   
1589196            성인                   일반           23.0          관리자   
1589311            성인                   일반           20.0          관리자   

        CRTFC_FLAG_NM  MESURE_DE  ... 라벨링_더미19  라벨링_더미20  라벨링_더미21  라벨링_더미22  \
4856              참가증   20120518  ...        0         0         0         0   
4967              참가증   20120521  ...        0         0         0         0   
5376              참가증   20120530  ...        0         0         0         0   
5407              참가증   20120531  ...        0         0         0         0   
5409              참가증   20120531  ...        0         0         0         0   
...               ...        ...  ...      ...       ...       ...       ...   
1589105           참가증   20220531  ...        0         0         0         1   
1589141           참가증   20220531  ...        0         0         0         1   
1589166           참가증   20220531  ...        0         0         0         1   
1589196           2등급   20220531  ...        1         0         0         0   
1589311           2등급   20220531  ...        1         0         0         0   

         라벨링_더미23  라벨링_더미24  라벨링_더미25  라벨링_더미26  라벨링_더미27  라벨링_더미28  
4856            0         0         0         0         0         0  
4967            0         0         0         0         0         0  
5376            0         0         0         0         0         0  
5407            0         0         0         0         0         0  
5409            0         0         0         0         0         0  
...           ...       ...       ...       ...       ...       ...  
1589105         0         0         0         1         0         0  
1589141         0         0         0         1         0         0  
1589166         0         0         0         1         0         0  
1589196         0         0         0         0         1         0  
1589311         0         0         0         0         1         0  

[75696 rows x 52 columns]

In [87]:
#csv파일로 저장
df_filtered.to_csv('/content/drive/MyDrive/데이터마이닝/데이터셋/main_exercises_ard.csv', index=False)
